roachtest: import/tpch/nodes=8 failed

[(roachtest).import/tpch/nodes=8 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1811809&tab=buildLog) on [master@5a3d0c9539a671f0e55b680d3021b18dde9d190d](https://github.com/cockroachdb/cockroach/commits/5a3d0c9539a671f0e55b680d3021b18dde9d190d):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200317-1811809/import/tpch/nodes=8/run_1
	cluster.go:2368,import.go:167,test_runner.go:753: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.registerImportTPCH.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/import.go:167
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - error with attached stack trace:
		    main.(*monitor).wait.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2412
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor task failed:
		  - pq: gs://cockroach-fixtures/tpch-csv/sf-100/lineitem.tbl.2: unexpected EOF
		    import failed
		    main.registerImportTPCH.func1.2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/import.go:163
		    main.(*monitor).Go.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2344
		    github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357

```

<details><summary>More</summary><p>


Artifacts: [/import/tpch/nodes=8](https://teamcity.cockroachdb.com/viewLog.html?buildId=1811809&tab=artifacts#/import/tpch/nodes=8)
Related:
- #45074 roachtest: import/tpch/nodes=8 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-44941](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-44941) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #44593 roachtest: import/tpch/nodes=8 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.2) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aimport%2Ftpch%2Fnodes%3D8.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).import/tpch/nodes=8 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1848230&tab=buildLog) on [master@c36f7cbfe08d48025540d0af62637a2fdd1f7d89](https://github.com/cockroachdb/cockroach/commits/c36f7cbfe08d48025540d0af62637a2fdd1f7d89):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200403-1848230/import/tpch/nodes=8/run_1
	cluster.go:2368,import.go:167,test_runner.go:753: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.registerImportTPCH.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/import.go:167
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - error with attached stack trace:
		    main.(*monitor).wait.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2412
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor task failed:
		  - pq: gs://cockroach-fixtures/tpch-csv/sf-100/lineitem.tbl.1: unexpected EOF
		    import failed
		    main.registerImportTPCH.func1.2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/import.go:163
		    main.(*monitor).Go.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2344
		    github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357

```

<details><summary>More</summary><p>


Artifacts: [/import/tpch/nodes=8](https://teamcity.cockroachdb.com/viewLog.html?buildId=1848230&tab=artifacts#/import/tpch/nodes=8)
Related:
- #45074 roachtest: import/tpch/nodes=8 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-44941](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-44941) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #44593 roachtest: import/tpch/nodes=8 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.2) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aimport%2Ftpch%2Fnodes%3D8.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).import/tpch/nodes=8 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1851668&tab=buildLog) on [master@82fec00c83d4bfe35b906264ccb568568cec15b7](https://github.com/cockroachdb/cockroach/commits/82fec00c83d4bfe35b906264ccb568568cec15b7):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200404-1851668/import/tpch/nodes=8/run_1
	test_runner.go:800: test timed out (8h0m0s)

	cluster.go:2368,import.go:167,test_runner.go:753: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2356
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2364
		    main.registerImportTPCH.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/import.go:167
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - error with attached stack trace:
		    main.(*monitor).wait.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2412
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor task failed:
		  - context canceled

```

<details><summary>More</summary><p>


Artifacts: [/import/tpch/nodes=8](https://teamcity.cockroachdb.com/viewLog.html?buildId=1851668&tab=artifacts#/import/tpch/nodes=8)
Related:
- #45074 roachtest: import/tpch/nodes=8 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-44941](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-44941) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #44593 roachtest: import/tpch/nodes=8 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.2) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aimport%2Ftpch%2Fnodes%3D8.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).import/tpch/nodes=8 failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1860251&tab=buildLog) on [master@10b3eeebba4ee1957b6f0c97fd64d972775e57ef](https://github.com/cockroachdb/cockroach/commits/10b3eeebba4ee1957b6f0c97fd64d972775e57ef):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200408-1860251/import/tpch/nodes=8/run_1
	cluster.go:2378,import.go:167,test_runner.go:753: error with attached stack trace:
		    main.(*monitor).WaitE
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2366
		    main.(*monitor).Wait
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2374
		    main.registerImportTPCH.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/import.go:167
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor failure:
		  - error with attached stack trace:
		    main.(*monitor).wait.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2422
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - monitor task failed:
		  - pq: gs://cockroach-fixtures/tpch-csv/sf-100/lineitem.tbl.6: Get https://storage.googleapis.com/cockroach-fixtures/tpch-csv/sf-100/lineitem.tbl.6?generation=1520636152541612: status code 500 trying to fetch http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token
		    import failed
		    main.registerImportTPCH.func1.2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/import.go:163
		    main.(*monitor).Go.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2354
		    github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/golang.org/x/sync/errgroup/errgroup.go:57
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357

```

<details><summary>More</summary><p>


Artifacts: [/import/tpch/nodes=8](https://teamcity.cockroachdb.com/viewLog.html?buildId=1860251&tab=artifacts#/import/tpch/nodes=8)
Related:
- #47199 roachtest: import/tpch/nodes=8 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #45074 roachtest: import/tpch/nodes=8 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-44941](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-44941) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

- #44593 roachtest: import/tpch/nodes=8 failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-19.2](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-19.2) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Aimport%2Ftpch%2Fnodes%3D8.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/storage/cloud/external_storage.go
+++ b/pkg/storage/cloud/external_storage.go
@@ -398,7 +398,9 @@ func delayedRetry(ctx context.Context, fn func() error) error {
 // happen if we didn't read from the connection too long due to e.g. load),
 // the same as unexpected eof errors.
 func isResumableHTTPError(err error) bool {
-	return errors.Is(err, io.ErrUnexpectedEOF) || sysutil.IsErrConnectionReset(err)
+	return errors.Is(err, io.ErrUnexpectedEOF) ||
+		sysutil.IsErrConnectionReset(err) ||
+		sysutil.IsErrConnectionRefused(err)
 }
 
 // Maximum number of times we can attempt to retry reading from external storage,

--- a/pkg/storage/cloud/gcs_storage.go
+++ b/pkg/storage/cloud/gcs_storage.go
@@ -167,32 +167,52 @@ type resumingGoogleStorageReader struct {
 	ctx    context.Context   // Reader context
 	bucket *gcs.BucketHandle // Bucket to read the data from
 	object string            // Object to read
-	data   io.ReadCloser     // Currently opened object data stream
+	data   *gcs.Reader       // Currently opened object data stream
 	pos    int64             // How much data was received so far
 }
 
 var _ io.ReadCloser = &resumingGoogleStorageReader{}
 
-func (r *resumingGoogleStorageReader) Read(p []byte) (n int, err error) {
-	for retries := 0; n == 0 && err == nil; retries++ {
-		n, err = r.data.Read(p)
-		r.pos += int64(n)
+func (r *resumingGoogleStorageReader) openStream() error {
+	return delayedRetry(r.ctx, func() error {
+		var readErr error
+		r.data, readErr = r.bucket.Object(r.object).NewRangeReader(r.ctx, r.pos, -1)
+		return readErr
+	})
+}
 
-		if err != nil && !errors.IsAny(err, io.EOF, io.ErrUnexpectedEOF) {
-			log.Errorf(r.ctx, "GCS:Read err: %s", err)
+func (r *resumingGoogleStorageReader) Read(p []byte) (int, error) {
+	var lastErr error
+
+	for retries := 0; lastErr == nil; retries++ {
+		if r.data == nil {
+			lastErr = r.openStream()
 		}
 
-		if isResumableHTTPError(err) {
-			log.Errorf(r.ctx, "GCS:Retry: error %s", err)
-			if retries > maxNoProgressReads {
-				err = errors.Wrap(err, "multiple Read calls return no data")
-				return
+		if lastErr == nil {
+			n, readErr := r.data.Read(p)
+			if readErr == nil {
+				r.pos += int64(n)
+				return n, nil
 			}
-			err = r.doRead()
+			lastErr = readErr
+		}
+
+		if !errors.IsAny(lastErr, io.EOF, io.ErrUnexpectedEOF) {
+			log.Errorf(r.ctx, "GCS:Read err: %s", lastErr)
+		}
+
+		if isResumableHTTPError(lastErr) {
+			if retries > maxNoProgressReads {
+				return 0, errors.Wrap(lastErr, "multiple Read calls return no data")
+			}
+			log.Errorf(r.ctx, "GCS:Retry: error %s", lastErr)
+			lastErr = nil
+			r.data = nil
 		}
 	}
 
-	return
+	return 0, lastErr
 }
 
 func (r *resumingGoogleStorageReader) Close() error {
@@ -202,29 +222,13 @@ func (r *resumingGoogleStorageReader) Close() error {
 	return nil
 }
 
-func (r *resumingGoogleStorageReader) doRead() error {
-	if r.data != nil {
-		if err := r.data.Close(); err != nil {
-			return err
-		}
-	}
-
-	return delayedRetry(r.ctx, func() error {
-		var readErr error
-		r.data, readErr = r.bucket.Object(r.object).NewRangeReader(r.ctx, r.pos, -1)
-		return readErr
-	})
-}
-
 func (g *gcsStorage) ReadFile(ctx context.Context, basename string) (io.ReadCloser, error) {
 	// https://github.com/cockroachdb/cockroach/issues/23859
-	var rc io.ReadCloser
-	err := delayedRetry(ctx, func() error {
-		var readErr error
-		rc, readErr = g.bucket.Object(path.Join(g.prefix, basename)).NewReader(ctx)
-		return readErr
-	})
-	return rc, err
+	return &resumingGoogleStorageReader{
+		ctx:    ctx,
+		bucket: g.bucket,
+		object: path.Join(g.prefix, basename),
+	}, nil
 }
 
 func (g *gcsStorage) ListFiles(ctx context.Context, patternSuffix string) ([]string, error) {

--- a/pkg/storage/cloud/gcs_storage_test.go
+++ b/pkg/storage/cloud/gcs_storage_test.go
@@ -1,0 +1,112 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package cloud
+
+import (
+	"context"
+	"io/ioutil"
+	"math/rand"
+	"net"
+	"net/http"
+	"os"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/cockroach/pkg/util/sysutil"
+	"github.com/stretchr/testify/require"
+)
+
+var econnrefused = &net.OpError{Err: &os.SyscallError{
+	Syscall: "test",
+	Err:     sysutil.ECONNREFUSED,
+}}
+
+var econnreset = &net.OpError{Err: &os.SyscallError{
+	Syscall: "test",
+	Err:     sysutil.ECONNRESET,
+}}
+
+type antagonisticDialer struct {
+	net.Dialer
+	rnd               *rand.Rand
+	numRepeatFailures *int
+}
+
+type antagonisticConn struct {
+	net.Conn
+	rnd               *rand.Rand
+	numRepeatFailures *int
+}
+
+func (d *antagonisticDialer) DialContext(
+	ctx context.Context, network, addr string,
+) (net.Conn, error) {
+	if network == "tcp" && addr == "storage.googleapis.com:443" {
+		if *d.numRepeatFailures < maxNoProgressReads && d.rnd.Int()%2 == 0 {
+			*(d.numRepeatFailures)++
+			return nil, econnrefused
+		}
+		c, err := d.Dialer.DialContext(ctx, network, addr)
+		if err != nil {
+			return nil, err
+		}
+
+		return &antagonisticConn{Conn: c, rnd: d.rnd, numRepeatFailures: d.numRepeatFailures}, nil
+	}
+	return d.Dialer.DialContext(ctx, network, addr)
+}
+
+func (c *antagonisticConn) Read(b []byte) (int, error) {
+	if *c.numRepeatFailures < maxNoProgressReads && c.rnd.Int()%2 == 0 {
+		*(c.numRepeatFailures)++
+		return 0, econnreset
+	}
+	return c.Conn.Read(b)
+}
+
+func TestAntagonisticRead(t *testing.T) {
+	if os.Getenv("GOOGLE_APPLICATION_CREDENTIALS") == "" {
+		// This test requires valid GS credential file.
+		return
+	}
+
+	rnd, _ := randutil.NewPseudoRand()
+	failures := 0
+	// Override DialContext implementation in http transport.
+	dialer := &antagonisticDialer{rnd: rnd, numRepeatFailures: &failures}
+
+	// Override transport to return antagonistic connection.
+	transport := http.DefaultTransport.(*http.Transport)
+	transport.DialContext =
+		func(ctx context.Context, network, addr string) (net.Conn, error) {
+			return dialer.DialContext(ctx, network, addr)
+		}
+	transport.DisableKeepAlives = true
+
+	defer func() {
+		transport.DialContext = nil
+		transport.DisableKeepAlives = false
+	}()
+
+	gsFile := "gs://cockroach-fixtures/tpch-csv/sf-1/region.tbl?AUTH=implicit"
+	conf, err := ExternalStorageConfFromURI(gsFile)
+	require.NoError(t, err)
+
+	s, err := MakeExternalStorage(
+		context.TODO(), conf, base.ExternalIOConfig{}, testSettings, nil)
+	require.NoError(t, err)
+	stream, err := s.ReadFile(context.TODO(), "")
+	require.NoError(t, err)
+	defer stream.Close()
+	_, err = ioutil.ReadAll(stream)
+	require.NoError(t, err)
+}

--- a/pkg/util/sysutil/sysutil.go
+++ b/pkg/util/sysutil/sysutil.go
@@ -15,11 +15,12 @@
 package sysutil
 
 import (
-	"net"
 	"os"
 	"os/exec"
 	"os/signal"
 	"syscall"
+
+	"github.com/cockroachdb/errors"
 )
 
 // Signal is syscall.Signal.
@@ -27,6 +28,12 @@ type Signal = syscall.Signal
 
 // Errno is syscall.Errno.
 type Errno = syscall.Errno
+
+// Exported syscall.Errno constants.
+const (
+	ECONNRESET   = syscall.ECONNRESET
+	ECONNREFUSED = syscall.ECONNREFUSED
+)
 
 // FSInfo describes a filesystem. It is returned by StatFS.
 type FSInfo struct {
@@ -64,10 +71,10 @@ func RefreshSignaledChan() <-chan os.Signal {
 // IsErrConnectionReset returns true if an
 // error is a "connection reset by peer" error.
 func IsErrConnectionReset(err error) bool {
-	if opErr, ok := err.(*net.OpError); ok {
-		if sysErr, ok := opErr.Err.(*os.SyscallError); ok {
-			return sysErr.Err == syscall.ECONNRESET
-		}
-	}
-	return false
+	return errors.Is(err, syscall.ECONNRESET)
+}
+
+// IsErrConnectionRefused returns true if an error is a "connection refused" error.
+func IsErrConnectionRefused(err error) bool {
+	return errors.Is(err, syscall.ECONNREFUSED)
 }

